picketlink/src/main/java/org/wildfly/extension/picketlink/federation/MigrateOperation.java

```diff
@@ -0,0 +1,339 @@
+                                        break;
+                                    }
+                                }
+                                context.getFailureDescription().set(ROOT_LOGGER.migrationFailed());
+                            }
+
+                            context.getResult().set(result);
+                        }
+                    });
+
+                }
+            }
+        }, MODEL);
+    }
+
+    private void describeLegacyPicketLinkFederationResources(OperationContext context, ModelNode legacyModelDescription) {
+        ModelNode describeLegacySubsystem = createOperation(GenericSubsystemDescribeHandler.DEFINITION, context.getCurrentAddress());
+        context.addStep(legacyModelDescription, describeLegacySubsystem, GenericSubsystemDescribeHandler.INSTANCE, MODEL, true);
+    }
+
+    /**
+     * Attempt to add the keycloak-saml extension. If it's already present, nothing is done.
+     */
+    private void addKeycloakSamlExtension(OperationContext context, Map<PathAddress, ModelNode> migrationOperations, boolean describe) {
+        Resource root = context.readResourceFromRoot(PathAddress.EMPTY_ADDRESS, false);
+        if (root.getChildrenNames(EXTENSION).contains(KEYCLOAK_SAML_EXTENSION)) {
+            return; // extension is already added, nothing to do
+        }
+        PathAddress extensionAddress = pathAddress(EXTENSION, KEYCLOAK_SAML_EXTENSION);
+        OperationEntry addEntry = context.getRootResourceRegistration().getOperationEntry(extensionAddress, ADD);
+        ModelNode addOperation = createAddOperation(extensionAddress);
+        addOperation.get(MODULE).set(KEYCLOAK_SAML_EXTENSION);
+        if (describe) {
+            migrationOperations.put(extensionAddress, addOperation);
+        } else {
+            context.addStep(context.getResult().get(extensionAddress.toString()), addOperation, addEntry.getOperationHandler(), MODEL);
+        }
+    }
+
+    private void transformResources(OperationContext context, final ModelNode legacyModelDescription, final Map<PathAddress, ModelNode> newAddOperations,
+                                    List<String> warnings) throws OperationFailedException {
+        for (ModelNode legacyAddOp : legacyModelDescription.get(RESULT).asList()) {
+            final ModelNode newAddOp = legacyAddOp.clone();
+            ModelNode legacyAddress = legacyAddOp.get(OP_ADDR);
+
+            ModelNode newAddress = transformAddress(legacyAddress.clone(), context);
+            if (! newAddress.isDefined()) {
+                continue;
+            }
+            newAddOp.get(OP_ADDR).set(newAddress);
+
+            PathAddress address = PathAddress.pathAddress(newAddress);
+
+            if (address.size() > 1) {
+                // non-empty subsystem configuration, fail for now
+                throw ROOT_LOGGER.cannotMigrateNonEmptyConfiguration();
+            }
+
+            if (newAddOp.isDefined()) {
```

Comment:
Why wouldn't it be defined at this point?

Reply:
Good catch, fixed.